src/EventStore.Core/Services/Transport/Grpc/PersistentSubscriptions.Read.cs

```diff
@@ -49,21 +51,26 @@ public partial class PersistentSubscriptions {
 
 			var subscriptionId = await enumerator.Started.ConfigureAwait(false);
 
-			var read = requestStream.ForEachAsync(HandleAckNack);
+			try {
+				var read = requestStream.ForEachAsync(HandleAckNack);
 
-			await responseStream.WriteAsync(new ReadResp {
-				SubscriptionConfirmation = new ReadResp.Types.SubscriptionConfirmation {
-					SubscriptionId = subscriptionId
-				}
-			}).ConfigureAwait(false);
-
-			while (await enumerator.MoveNextAsync().ConfigureAwait(false)) {
 				await responseStream.WriteAsync(new ReadResp {
-					Event = ConvertToReadEvent(enumerator.Current)
+					SubscriptionConfirmation = new ReadResp.Types.SubscriptionConfirmation {
+						SubscriptionId = subscriptionId
+					}
 				}).ConfigureAwait(false);
-			}
 
-			await read.ConfigureAwait(false);
+				while (await enumerator.MoveNextAsync().ConfigureAwait(false)) {
+					await responseStream.WriteAsync(new ReadResp {
+						Event = ConvertToReadEvent(enumerator.Current)
+					}).ConfigureAwait(false);
+				}
+
+				await read.ConfigureAwait(false);
+			} catch (IOException) {
+				Log.Information("Subscription {correlationId} to {subscriptionId} disposed. The request stream was closed.", correlationId, subscriptionId);
```

Comment:
Should this be a warning?

Reply:
We get this exception under perfectly normal circumstances, such as disposing a subscription on the client or a client disconnecting. So if we changed this to a warning, then we would log out warnings for normal operations, though there may be cases when this is a legitimate error.

The following are the exceptions we get for disposing the subscription and disconnecting the client. We could try something like a pattern match for these specific exceptions, but to me they look generic enough to be caused by other things as well.

```[ 9768,29,09:27:46.148,ERR] Error reading message.
System.IO.IOException: The client reset the request stream.
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2MessageBody.ReadAsync(CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.PipeExtensions.ReadStreamMessageAsync[T](PipeReader input, HttpContextServerCallContext serverCallContext, Func`2 deserializer, CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.HttpContextStreamReader`1.<MoveNext>g__MoveNextAsync|10_0(ValueTask`1 readStreamTask)
   at EventStore.Core.Services.Transport.Grpc.AsyncStreamExtensions.ForEachAsync[T](IAsyncStreamReader`1 streamReader, Func`2 asyncAction) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\AsyncStreamExtensions.cs:line 34
   at EventStore.Core.Services.Transport.Grpc.PersistentSubscriptions.Read(IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream, ServerCallContext context) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\PersistentSubscriptions.Read.cs:line 66
   at EventStore.Core.Services.Transport.Grpc.PersistentSubscriptions.Read(IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream, ServerCallContext context) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\PersistentSubscriptions.Read.cs:line 66
   at Grpc.Shared.Server.DuplexStreamingServerMethodInvoker`3.Invoke(HttpContext httpContext, ServerCallContext serverCallContext, IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream)
   at Grpc.Shared.Server.DuplexStreamingServerMethodInvoker`3.Invoke(HttpContext httpContext, ServerCallContext serverCallContext, IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream)
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerCallHandlerBase`3.<HandleCallAsync>g__AwaitHandleCall|8_0(HttpContextServerCallContext serverCallContext, Method`2 method, Task handleCall)
[ 9768,29,09:27:46.155,ERR] Error when executing service method '"Read"'.
System.IO.IOException: The client reset the request stream.
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2MessageBody.ReadAsync(CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.PipeExtensions.ReadStreamMessageAsync[T](PipeReader input, HttpContextServerCallContext serverCallContext, Func`2 deserializer, CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.HttpContextStreamReader`1.<MoveNext>g__MoveNextAsync|10_0(ValueTask`1 readStreamTask)
   at EventStore.Core.Services.Transport.Grpc.AsyncStreamExtensions.ForEachAsync[T](IAsyncStreamReader`1 streamReader, Func`2 asyncAction) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\AsyncStreamExtensions.cs:line 34
   at EventStore.Core.Services.Transport.Grpc.PersistentSubscriptions.Read(IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream, ServerCallContext context) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\PersistentSubscriptions.Read.cs:line 66
   at EventStore.Core.Services.Transport.Grpc.PersistentSubscriptions.Read(IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream, ServerCallContext context) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\PersistentSubscriptions.Read.cs:line 66
   at Grpc.Shared.Server.DuplexStreamingServerMethodInvoker`3.Invoke(HttpContext httpContext, ServerCallContext serverCallContext, IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream)
   at Grpc.Shared.Server.DuplexStreamingServerMethodInvoker`3.Invoke(HttpContext httpContext, ServerCallContext serverCallContext, IAsyncStreamReader`1 requestStream, IServerStreamWriter`1 responseStream)
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerCallHandlerBase`3.<HandleCallAsync>g__AwaitHandleCall|8_0(HttpContextServerCallContext serverCallContext, Method`2 method, Task handleCall)

[ 9768,40,09:29:03.508,ERR] Error reading message.
System.IO.IOException: The request stream was aborted.
 ---> Microsoft.AspNetCore.Connections.ConnectionAbortedException: The HTTP/2 connection faulted.
 ---> Microsoft.AspNetCore.Connections.ConnectionResetException: An existing connection was forcibly closed by the remote host.
 ---> System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host.
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.<GetResult>g__ThrowSocketException|7_0(SocketError e)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
   --- End of inner exception stack trace ---
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.DuplexPipeStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken)
   at System.Net.Security.SslStream.<FillBufferAsync>g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial)
   at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
   at System.IO.Pipelines.StreamPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ReadInputAsync()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2MessageBody.ReadAsync(CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.PipeExtensions.ReadStreamMessageAsync[T](PipeReader input, HttpContextServerCallContext serverCallContext, Func`2 deserializer, CancellationToken cancellationToken)
   at Grpc.AspNetCore.Server.Internal.HttpContextStreamReader`1.<MoveNext>g__MoveNextAsync|10_0(ValueTask`1 readStreamTask)
   at EventStore.Core.Services.Transport.Grpc.AsyncStreamExtensions.ForEachAsync[T](IAsyncStreamReader`1 streamReader, Func`2 asyncAction) in C:\Users\hayle\code\EventStore\src\EventStore.Core\Services\Transport\Grpc\AsyncStreamExtensions.cs:line 34
```

Reply:
As a side note, we could include the description of the exception (`he client reset the request stream` or `The request stream was aborted`) in the log message